File: src/Microservices/AdvertisementsMicroservice/AdvertisementsMicroservice.BLL/Services/AdvertisementsService.cs
using AdvertisementsMicroservice.BLL.Interfaces;
using AdvertisementsMicroservice.DAL.Interfaces;
using AdvertisingAgency.Contracts.Requests;
using AdvertisingAgency.Contracts.Responses;

namespace AdvertisementsMicroservice.BLL.Services
{
    public class AdvertisementsService : IAdvertisementsService
    {
        private readonly IBannerRepository _bannerRepository;
        private readonly IUserBannerRepository _userBannerRepository;
        private readonly IStatusRepository _statusRepository;

        public AdvertisementsService(IBannerRepository bannerRepository, IUserBannerRepository userBannerRepository, IStatusRepository statusRepository)
        {
            _bannerRepository = bannerRepository;
            _userBannerRepository = userBannerRepository;
            _statusRepository = statusRepository;
        }

        public async Task<MessageResponse> AddBanner(AddBannerRequest addBannerRequest)
        {
            var bannerId = await _bannerRepository.AddBannerAndGetBannerId(
                addBannerRequest.Title, 
                addBannerRequest.SubTitle, 
                addBannerRequest.Description, 
                addBannerRequest.LinkToBrowserPage,
                addBannerRequest.ImageUrl);

            await _userBannerRepository.AddUserBanner(addBannerRequest.UserId, bannerId);

            return new MessageResponse { Message = "Banner was added successful" };
        }

        public async Task<GetAllBannersResponse> GetAllBannersByUserId(GetAllBannersByUserIdRequest getAllBannersByUserIdRequest)
        {
            var userId = Guid.Parse(getAllBannersByUserIdRequest.UserId);
            var userBanners = await _userBannerRepository.GetAllUserBannersByUserId(userId);

            var response = new GetAllBannersResponse();

            foreach (var item in userBanners)
            {
                var banner = await _bannerRepository.GetBannerByIdAnyStatus(item.BannerId.ToString());
                if (banner != null)
                {
                    response.BannerList.Add(new BannerResponse
                    {
                        UserId = item.UserId.ToString(),
                        BannerId = item.BannerId.ToString(),
                        Title = banner.Title,
                        SubTitle = banner.SubTitle,
                        Description = banner.Description,
                        LinkToBrowserPage = banner.LinkToBrowserPage,
                        ReleaseDate = banner.ReleaseDate.ToShortDateString(),
                        Status = await _statusRepository.GetStatusNameById(banner.StatusId),
                        PhotoUrl = banner.PhotoUrl,
                        Comment = banner.Comment
                    });
                }
            }

            return response;
        }

        public async Task<GetAllBannersResponse> GetAllBanners()
        {
            var userBanners = await _userBannerRepository.GetAllUserBanners();

            var response = new GetAllBannersResponse();

            foreach (var item in userBanners)
            {
                var banner = await _bannerRepository.GetBannerById(item.BannerId.ToString());
                if (banner != null)
                {
                    response.BannerList.Add(new BannerResponse
                    {
                        UserId = item.UserId.ToString(),
                        BannerId = item.BannerId.ToString(),
                        Title = banner.Title,
                        SubTitle = banner.SubTitle,
                        Description = banner.Description,
                        LinkToBrowserPage = banner.LinkToBrowserPage,
                        ReleaseDate = banner.ReleaseDate.ToShortDateString(),
                        Status = await _statusRepository.GetStatusNameById(banner.StatusId),
                        PhotoUrl = banner.PhotoUrl,
                        Comment = banner.Comment
                    });
                }
            }

            return response;
        }

        public async Task<MessageResponse> DeleteBanner(DeleteBannerRequest deleteBannerRequest)
        {
            await _userBannerRepository.DeleteUserBanner(deleteBannerRequest.UserId, deleteBannerRequest.BannerId);
            return new MessageResponse { Message = "Delete successful!" };
        }

        public async Task<MessageResponse> AddToQueueToCheck(AddToQueueToCheckRequest addToQueueToCheckRequest)
        {
            await _bannerRepository.AddBannerToQueueToCheck(addToQueueToCheckRequest.BannerId);
            return new MessageResponse { Message = "Successful added to queue to check" };
        }

        public async Task<GetAllBannersResponse> GetAllBannersWhereStatusInQueueToCheck()
        {
            var userBanners = await _userBannerRepository.GetAllUserBanners();

            var response = new GetAllBannersResponse();

            foreach (var item in userBanners)
            {
                var banner = await _bannerRepository.GetBannerByIdWhereToQueueToCheck(item.BannerId.ToString());
                if (banner != null)
                {
                    response.BannerList.Add(new BannerResponse
                    {
                        UserId = item.UserId.ToString(),
                        BannerId = item.BannerId.ToString(),
                        Title = banner.Title,
                        SubTitle = banner.SubTitle,
                        Description = banner.Description,
                        LinkToBrowserPage = banner.LinkToBrowserPage,
                        ReleaseDate = banner.ReleaseDate.ToShortDateString(),
                        Status = await _statusRepository.GetStatusNameById(banner.StatusId),
                        PhotoUrl = banner.PhotoUrl,
                        Comment = banner.Comment
                    });
                }
            }

            return response;
        }

        public async Task<GetAllBannersResponse> GetAllBannersWhereStatusCheckSuccessful()
        {
            var userBanners = await _userBannerRepository.GetAllUserBanners();

            var response = new GetAllBannersResponse();

            foreach (var item in userBanners)
            {
                var banner = await _bannerRepository.GetBannerByIdWhereCheckSuccessful(item.BannerId.ToString());
                if (banner != null)
                {
                    response.BannerList.Add(new BannerResponse
                    {
                        UserId = item.UserId.ToString(),
                        BannerId = item.BannerId.ToString(),
                        Title = banner.Title,
                        SubTitle = banner.SubTitle,
                        Description = banner.Description,
                        LinkToBrowserPage = banner.LinkToBrowserPage,
                        ReleaseDate = banner.ReleaseDate.ToShortDateString(),
                        Status = await _statusRepository.GetStatusNameById(banner.StatusId),
                        PhotoUrl = banner.PhotoUrl,
                        Comment = banner.Comment
                    });
                }
            }

            return response;
        }

        public async Task<MessageResponse> SetStatusCheckSuccessful(SetStatusCheckSuccessfulRequest setStatusCheckSuccessfulRequest)
        {
            await _bannerRepository.SetStatusCheckSuccessful(setStatusCheckSuccessfulRequest.BannerId);
            return new MessageResponse { Message = "Successful update status to Check Successful" };
        }

        public async Task<MessageResponse> SetStatusCheckNotSuccessful(SetStatusCheckNotSuccessfulRequest setStatusCheckSuccessfulRequest)
        {
            await _bannerRepository.SetStatusCheckNotSuccessful(setStatusCheckSuccessfulRequest.BannerId, setStatusCheckSuccessfulRequest.Comment);
            return new MessageResponse { Message = "Successful update status to Check Not Successful" };
        }

        public async Task<MessageResponse> SetStatusReleased(SetStatusReleasedRequest setStatusReleasedRequest)
        {
            await _bannerRepository.SetStatusReleased(setStatusReleasedRequest.BannerId);
            return new MessageResponse { Message = "Successful update status to Released" };
        }

        public async Task<MessageResponse> SetStatusReleasePlanned(SetStatusReleasePlannedRequest setStatusReleasePlannedRequest)
        {
            await _bannerRepository.SetStatusReleasePlanned(setStatusReleasePlannedRequest.BannerId, setStatusReleasePlannedRequest.ReleaseDate);
            return new MessageResponse { Message = "Successful update status to Released Planned" };
        }

        public async Task<MessageResponse> UpdateBanner(UpdateBannerRequest updateAdvertisementRequest)
        {
            await _bannerRepository.UpdateBanner(
                updateAdvertisementRequest.BannerId,
                updateAdvertisementRequest.Title,
                updateAdvertisementRequest.SubTitle,
                updateAdvertisementRequest.Description,
                updateAdvertisementRequest.LinkToBrowserPage,
                updateAdvertisementRequest.PhotoUrl);
            return new MessageResponse { Message = "Update advertisement successfull" };
        }
    }
}

File: AngularClientApp/src/app/services/advertisements.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { AddAdvertisementCredentials } from '../credentials/add-advertisement-credentials';
import { RejectCredentials } from '../credentials/reject-credentials';
import { ReleasePlannedCredentials } from '../credentials/release-planned-credentials';
import { UpdateAdvertisementCredentials } from '../credentials/update-advertisement-credentials';
import { Advertisement } from '../models/advertisement';
import { UpdateAdvertisementResponse } from '../models/update-advertisement-reponse';

@Injectable({
  providedIn: 'root'
})
export class AdvertisementsService {
  baseUrl = "https://localhost:56782/api";

  httpOptions = {
    headers: new HttpHeaders({'Content-Type': 'application/json'})
  }

  constructor(private http: HttpClient) { }

  getAllBanners(message: string){
    return this.http.get<Advertisement[]>(this.baseUrl + `/Advertisements/getAllBanners?message=${message}`);
  }

  getAllBannersByUserId(userId: string) {
    return this.http.get<Advertisement[]>(this.baseUrl + `/Advertisements/getAllBannersByUserId?userId=${userId}`);
  }

  addBanner(credentials: AddAdvertisementCredentials){
    return this.http.post<string>(this.baseUrl + '/Advertisements/addBanner', credentials, this.httpOptions);
  }

  deleteBannerByUserIdAndBannerId(userId: string, bannerId: string) {
    return this.http.delete(this.baseUrl + `/Advertisements/deleteBanner?userId=${userId}&&bannerId=${bannerId}`);
  }

  addToQueueToCheck(bannerId: string){
    return this.http.put(this.baseUrl + '/Advertisements/addToQueueToCheck', { 'BannerId': bannerId }, this.httpOptions);
  }

  getAllBannersWhereStatusInQueueToCheck() {
    return this.http.get<Advertisement[]>(this.baseUrl + '/Advertisements/getAllBannersWhereStatusInQueueToCheck');
  }

  setStatusCheckSuccessfull(bannerId: string) {
    return this.http.put(this.baseUrl + '/Advertisements/setStatusCheckSuccessfull', { 'BannerId': bannerId }, this.httpOptions);
  }

  setStatusCheckNotSuccessfull(credentials: RejectCredentials) {
    return this.http.put(this.baseUrl + '/Advertisements/setStatusCheckNotSuccessful', { 'BannerId': credentials.bannerId, 'Comment': credentials.comment }, this.httpOptions);
  }

  setStatusReleased(bannerId: string){
    return this.http.put(this.baseUrl + '/Advertisements/setStatusReleased', { 'BannerId': bannerId }, this.httpOptions);
  }

  setStatusReleasePlanned(credentials: ReleasePlannedCredentials) {
    return this.http.put(this.baseUrl + '/Advertisements/setStatusReleasePlanned', { 'BannerId': credentials.bannerId, 'ReleaseDate': credentials.releaseDate.toString() }, this.httpOptions);
  }

  getAllBannersWhereStatusCheckSuccessful() {
    return this.http.get<Advertisement[]>(this.baseUrl + '/Advertisements/getAllBannersWhereStatusCheckSuccessful');
  }

  updateBanner(credentials: UpdateAdvertisementCredentials){
    return this.http.put<UpdateAdvertisementResponse>(this.baseUrl + '/Advertisements/updateBanner', credentials, this.httpOptions);
  }
}